src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -725,27 +736,480 @@ mod commutation_tests {
+                for acc in Access2::exhaustive() {
+                    let Ok(new_source) = source.perform_access2(acc) else { continue; };
+                    let Ok(new_target) = target.perform_access2(acc) else { return true; };
+                    if new_source.trivially_indistinguishable(&new_target) { continue; }
+                    states.push((new_source, new_target));
+                }
+                if source.x.prot {
+                    let Ok(new_source) = source.release_x() else { continue; };
+                    let Ok(new_target) = target.release_x() else { return true; };
+                    if new_source.trivially_indistinguishable(&new_target) { continue; }
+                    states.push((new_source, new_target));
+                }
+                if source.y.prot {
+                    let Ok(new_source) = source.release_y() else { continue; };
+                    let Ok(new_target) = target.release_y() else { return true; };
+                    if new_source.trivially_indistinguishable(&new_target) { continue; }
+                    states.push((new_source, new_target));
+                }
+            }
+            false
+        }
+    }
+
+    #[test]
+    #[should_panic]
+    // This is why `Reserved -> Frozen` on foreign read for protected references
+    // prevents the insertion of spurious reads: the transition can cause UB in the target
+    // later down the line.
+    fn reserved_frozen_protected_distinguishable() {
+        let source = LocStateProt2 {
+            x: LocStateProt {
+                state: LocationState::new(Permission::new_frozen()).with_access(),
+                prot: true,
+            },
+            y: LocStateProt {
+                state: LocationState::new(Permission::new_reserved(false)),
+                prot: true,
+            },
+        };
+        let acc = Access2 { rel2: AccessRel2::ChildX, kind: AccessKind::Read };
+        let target = source.perform_access2(acc).unwrap();
+        assert!(source.y.state.permission.is_reserved(None));
+        assert!(target.y.state.permission.is_frozen());
+        assert!(!source.distinguishable(&target))
+    }
+
+    #[test]
+    #[should_panic]
+    /// This test in a generalization of `reserved_frozen_protected_distinguishable`
+    /// that more exhaustively explores the feasability of inserting a spurious read.
+    /// For a number of interleavings of different pieces of code involving
+    /// two mutually foreign tags, we test if a spurious read through one
+    /// changes the outcome (introduces UB in the target when there is none
+    /// in the source).
+    fn spurious_read() {
+        // The general code pattern being exhaustively tested is
+        //     [thread 1]          || [thread 2]
+        //     retag x (protect)   || retag y (protect/noprotect)
+        //                    arbitrary code
+        //                         || optional return
+        //     <spurious read x>   ||
```

Comment:
Needs to explain that x must still be protected when we do the spurious read, that's why x is fixed as protected and there is no optional return here. (There is an optional return in the distinguishability test though.)